src/backend/dx12/src/device.rs

```diff
@@ -1792,250 +1790,265 @@ impl d::Device<B> for Device {
-            // The DepthBoundsTestEnable option isn't available in the original D3D12_GRAPHICS_PIPELINE_STATE_DESC struct.
-            // Instead, we must use the newer subobject stream method.
-            let (device2, hr) = self.raw.cast::<d3d12::ID3D12Device2>();
-            if winerror::SUCCEEDED(hr) {
-                let mut pss_stream = GraphicsPipelineStateSubobjectStream::new(&pso_desc, true);
-                let pss_desc = d3d12::D3D12_PIPELINE_STATE_STREAM_DESC {
-                    SizeInBytes: mem::size_of_val(&pss_stream),
-                    pPipelineStateSubobjectStream: &mut pss_stream as *mut _ as _,
+                let destroy_shader = |shader: ShaderBc| {
+                    if let ShaderBc::Owned(bc) = shader {
+                        bc.destroy();
+                    }
                 };
-                device2.CreatePipelineState(
-                    &pss_desc,
-                    &d3d12::ID3D12PipelineState::uuidof(),
-                    pipeline.mut_void(),
-                )
-            } else {
-                hr
-            }
-        } else {
-            self.raw.clone().CreateGraphicsPipelineState(
-                &pso_desc,
-                &d3d12::ID3D12PipelineState::uuidof(),
-                pipeline.mut_void(),
-            )
-        };
 
-        let destroy_shader = |shader: ShaderBc| {
-            if let ShaderBc::Owned(bc) = shader {
-                bc.destroy();
-            }
-        };
+                destroy_shader(vs);
+                destroy_shader(ps);
+                destroy_shader(gs);
+                destroy_shader(hs);
+                destroy_shader(ds);
 
-        destroy_shader(vs);
-        destroy_shader(ps);
-        destroy_shader(gs);
-        destroy_shader(hs);
-        destroy_shader(ds);
+                if winerror::SUCCEEDED(hr) {
+                    let mut baked_states = desc.baked_states.clone();
+                    if !desc.depth_stencil.depth_bounds {
+                        baked_states.depth_bounds = None;
+                    }
 
-        if winerror::SUCCEEDED(hr) {
-            let mut baked_states = desc.baked_states.clone();
-            if !desc.depth_stencil.depth_bounds {
-                baked_states.depth_bounds = None;
+                    Ok(r::GraphicsPipeline {
+                        raw: pipeline,
+                        signature: desc.layout.raw,
+                        num_parameter_slots: desc.layout.num_parameter_slots,
+                        topology,
+                        constants: desc.layout.constants.clone(),
+                        vertex_bindings,
+                        baked_states,
+                    })
+                } else {
+                    Err(pso::CreationError::Other)
+                }
+            }
+            pso::PrimitiveAssembler::Mesh { .. } => {
+                // TODO: Implement mesh shaders for DX12
+                Err(pso::CreationError::UnsupportedPipeline)
```

Comment:
it's somewhat unfortunate to indent the whole multi-page block above because of this. Maybe we can have it as early return for now?

Reply:
So the new version is the best I could do without moving large indented blocks inside match and also satisfy Rust borrow checker. 

There are probably a lot of variants It could be rewritten. I might need some further advice on this if this is not satisfactory.